server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -720,25 +742,109 @@ private static ActionListener<SearchResponse> createCCSListener(
             countDown,
             skippedClusters,
             exceptions,
+            clusters.getCluster(clusterAlias),
             originalListener
         ) {
             @Override
             void innerOnResponse(SearchResponse searchResponse) {
+                // TODO: in CCS fail fast ticket we may need to fail the query if the cluster gets marked as FAILED
+                ccsClusterInfoUpdate(searchResponse, cluster, skipUnavailable);
                 searchResponseMerger.add(searchResponse);
             }
 
             @Override
             SearchResponse createFinalResponse() {
-                SearchResponse.Clusters clusters = new SearchResponse.Clusters(
-                    totalClusters,
-                    searchResponseMerger.numResponses(),
-                    skippedClusters.get()
-                );
                 return searchResponseMerger.getMergedResponse(clusters);
             }
         };
     }
 
+    /**
+     * Creates a new Cluster object using the {@link ShardSearchFailure} info and skip_unavailable
+     * flag to set Status. The new Cluster object is swapped into the clusterRef {@link AtomicReference}.
+     */
+    static void ccsClusterInfoUpdate(
+        ShardSearchFailure failure,
+        AtomicReference<SearchResponse.Cluster> clusterRef,
+        boolean skipUnavailable
+    ) {
+        SearchResponse.Cluster.Status status;
+        if (skipUnavailable) {
+            status = SearchResponse.Cluster.Status.SKIPPED;
+        } else {
+            status = SearchResponse.Cluster.Status.FAILED;
+        }
+        boolean swapped;
+        do {
+            SearchResponse.Cluster orig = clusterRef.get();
+            String clusterAlias = orig.getClusterAlias();
+            List<ShardSearchFailure> failures;
+            if (orig.getFailures() != null) {
+                failures = new ArrayList<>(orig.getFailures());
+            } else {
+                failures = new ArrayList<>(1);
+            }
+            failures.add(failure);
+            String indexExpression = orig.getIndexExpression();
+            SearchResponse.Cluster updated = new SearchResponse.Cluster(clusterAlias, indexExpression, status, failures);
+            swapped = clusterRef.compareAndSet(orig, updated);
+        } while (swapped == false);
```

Comment:
This concurrency model was primarily designed with MRT=false in mind but I wanted one model for both systems so I implemented it here as well. 

However, in my next PR (CCS fail fast), this model is also now useful for MRT=true, because other Cluster objects will have their status updated to `CANCELLED` (a new status not in this PR) in the callback thread of the Cluster that is failing the search, so you could then have one thread trying to update the Cluster object to have, say, `SUCCESSFUL` status, and the "cancellation" thread updating that same Cluster reference to `CANCELLED` status at the same time, so this CAS concurrency control model will be useful to 1) avoid losing state added by another thread and 2) allow a sticky state machine (once the Cluster is moved to CANCELLED it should not be changed).